sweepbatcher/sweep_batch.go

```diff
@@ -976,6 +1006,340 @@ func (b *batch) publishBatchCoop(ctx context.Context) (btcutil.Amount,
+				err)
+		}
+
+		// Create and store the previous outpoint for this sweep.
+		prevOut := &wire.TxOut{
+			Value:    int64(sweep.value),
+			PkScript: sweep.htlc.PkScript,
+		}
+		prevOutsList = append(prevOutsList, prevOut)
+
+		// Create and store the sign descriptor for this sweep.
+		signDesc := lndclient.SignDescriptor{
+			WitnessScript: sweep.htlc.SuccessScript(),
+			Output:        prevOut,
+			HashType:      sweep.htlc.SigHash(),
+			InputIndex:    i,
+			KeyDesc: keychain.KeyDescriptor{
+				PubKey: key,
+			},
+		}
+
+		if sweep.htlc.Version == swap.HtlcV3 {
+			signDesc.SignMethod = input.TaprootScriptSpendSignMethod
+		}
+
+		signDescs = append(signDescs, &signDesc)
+	}
+
+	// Sanity checks.
+	if len(signDescs) != nonCoopInputs {
+		// This must not happen by construction.
+		return 0, fmt.Errorf("unexpected size of signDescs: %d != %d",
+			len(signDescs), nonCoopInputs)
+	}
+	if len(prevOutsList) != nonCoopInputs {
+		// This must not happen by construction.
+		return 0, fmt.Errorf("unexpected size of prevOutsList: "+
+			"%d != %d", len(prevOutsList), nonCoopInputs)
+	}
+
+	var rawSigs [][]byte
+	if nonCoopInputs > 0 {
+		// Produce the signatures for our inputs using sign descriptors.
+		var err error
+		rawSigs, err = b.signerClient.SignOutputRaw(
+			ctx, tx, signDescs, prevOutsList,
+		)
+		if err != nil {
+			return 0, fmt.Errorf("signerClient.SignOutputRaw "+
+				"failed: %w", err)
+		}
+	}
+
+	// Sanity checks.
+	if len(rawSigs) != nonCoopInputs {
+		// This must not happen by construction.
+		return 0, fmt.Errorf("unexpected size of rawSigs: %d != %d",
+			len(rawSigs), nonCoopInputs)
+	}
+
+	// Generate success witnesses for non-cooperative sweeps.
+	sigIndex := 0
+	for i, sweep := range sweeps {
+		// Skip cooperative sweeps.
+		if !sweep.nonCoopHint && !sweep.coopFailed {
+			continue
+		}
+
+		witness, err := sweep.htlc.GenSuccessWitness(
+			rawSigs[sigIndex], sweep.preimage,
+		)
+		if err != nil {
+			return 0, fmt.Errorf("sweep.htlc.GenSuccessWitness "+
+				"failed: %w", err)
+		}
+		sigIndex++
+
+		// Add the success witness to our batch transaction's inputs.
+		tx.TxIn[i].Witness = witness
+	}
+
+	// Log transaction's details.
+	txHash := tx.TxHash()
+	b.log.Infof("attempting to publish mixed tx=%v with feerate=%v, "+
+		"weight=%v, feeForWeight=%v, fee=%v, sweeps=%d (%d cooperative"+
+		" + %d non-cooperative), destAddr=%s",
+		txHash, b.rbfCache.FeeRate, weight, feeForWeight, fee,
+		len(tx.TxIn), coopInputs, nonCoopInputs, address)
+
+	b.debugLogTx("serialized mixed batch", tx)
+
+	// Make sure tx weight matches the expected value.
+	realWeight := lntypes.WeightUnit(
+		blockchain.GetTransactionWeight(btcutil.NewTx(tx)),
+	)
+	if realWeight != weight {
+		b.log.Warnf("actual weight of tx %v is %v, estimated as %d",
+			txHash, realWeight, weight)
+	}
+
+	// Publish the transaction.
+	err := b.wallet.PublishTransaction(
+		ctx, tx, labels.LoopOutBatchSweepSuccess(b.id),
+	)
+	if err != nil {
```

Comment:
Ideally I think we'd want to filter out the case where replace fails (but the tx itself is valid) so we don't end up ditching the coops sweeps in favor of revealing the scripts etc.

Reply:
Done. See the comment above, about "signSuccess".

Reply:
I think this requires checking for publishing errors like `output already spent` I think.